sdk/core/azure-core/src/http/policy_retry.cpp

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <http/policy.hpp>
+
+#include <algorithm>
+#include <cstdlib>
+#include <limits>
+#include <thread>
+
+using namespace Azure::Core::Http;
+
+namespace {
+decltype(RetryOptions::RetryDelay) ShouldRetry(
```

Comment:
I feel that `decltype` and `auto` everywhere makes this file really hard to read.

Reply:
Do you know any open source projects where people use `decltype` so frequently or do you have other justification that this is a good coding style or practice?

Reply:
I've been writing C++ for many years, and I personally think this is very uncommon.

Reply:
I would use it less in public headers, but this is unnamed namespace in the implementation, and I like how we can change the underlying type in struct, and it gets picked up by implementation without any mismatch. We can change milliseconds from int64 to int32, and there's no need to see what are the places now that needs to be updated.

Reply:
Yes, but when i'm reviewing this piece of code, i have difficulty figuring out the return type, the parameter type, which makes it harder for me to understand the code. I guess other people have the same kinds of problems.

Reply:
I agree with Jinming on this. Code understandability and maintainability is important for the sustained life of our code base. decltype inserts a level of indirection into the source code which hurts the understandability. Also, if the type were to change, it is probably good to look at all the code using it to ensure that the type change doesn't break anything.

Reply:
I reduced the usage of decltype, using typedefs.